The ClassCastException for org.springframework.http.HttpHeaders instance in feign.Contract

Hello,
I updated my project to `Spring Boot 2`. `Feign` library was migrated from version `9.5.0` to `9.5.1` in `Spring Boot 2`. And now my `feign` version is `9.5.1` and my spring boot app failed.

I'm using `org.springframework.http.HttpHeaders` class and have `ClassCastException`. 
If change `HttpHeaders` to `java.util.Map` everything is ok.

Stacktrace:
```
Caused by: java.lang.ClassCastException: java.lang.Class cannot be cast to java.lang.reflect.ParameterizedType
	at feign.Contract$BaseContract.checkMapString(Contract.java:137) ~[feign-core-9.5.1.jar:na]
	at feign.Contract$BaseContract.parseAndValidateMetadata(Contract.java:124) ~[feign-core-9.5.1.jar:na]
	at org.springframework.cloud.openfeign.support.SpringMvcContract.parseAndValidateMetadata(SpringMvcContract.java:133) ~[spring-cloud-openfeign-core-2.0.0.M1.jar:2.0.0.M1]
	at feign.Contract$BaseContract.parseAndValidatateMetadata(Contract.java:66) ~[feign-core-9.5.1.jar:na]
	at feign.hystrix.HystrixDelegatingContract.parseAndValidatateMetadata(HystrixDelegatingContract.java:34) ~[feign-hystrix-9.5.1.jar:na]
	at feign.ReflectiveFeign$ParseHandlersByName.apply(ReflectiveFeign.java:146) ~[feign-core-9.5.1.jar:na]
	at feign.ReflectiveFeign.newInstance(ReflectiveFeign.java:53) ~[feign-core-9.5.1.jar:na]
	at feign.Feign$Builder.target(Feign.java:218) ~[feign-core-9.5.1.jar:na]
```

The problem appeared from that commit https://github.com/OpenFeign/feign/commit/a029afd8666b9a8ddeaecd1599e10b7b55f43bb4
```java
feign.Contract.java:
// 137 line
Type[] parameterTypes = ((ParameterizedType) genericType).getActualTypeArguments();
```

My app:
```java
public interface GraphQLClient {
    @RequestMapping(value = "/graphql/d", method = RequestMethod.POST)
    ExecutionResultWrapper execute(@RequestBody QueryWrapper queryWrapper, @RequestHeader HttpHeaders httpHeaders);
}
```

```java
import own.DefaultFeignConfiguration;
import own.CustomErrorDecoder;
import feign.Client;
import feign.Target;
import feign.hystrix.HystrixFeign;
import feign.slf4j.Slf4jLogger;
import org.springframework.beans.factory.ObjectFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.http.HttpMessageConverters;
import org.springframework.cloud.netflix.ribbon.SpringClientFactory;
import org.springframework.cloud.openfeign.ribbon.CachingSpringLoadBalancerFactory;
import org.springframework.cloud.openfeign.ribbon.LoadBalancerFeignClient;
import org.springframework.cloud.openfeign.support.ResponseEntityDecoder;
import org.springframework.cloud.openfeign.support.SpringDecoder;
import org.springframework.cloud.openfeign.support.SpringEncoder;
import org.springframework.cloud.openfeign.support.SpringMvcContract;
import org.springframework.stereotype.Component;

@Component
public class ClientBuilder {
    // ...
    public <T> T buildClient(Class<T> targetInterface, String prefix, String url) {
        Target<T> custom = new Target.HardCodedTarget<>(targetInterface, url, prefix + url);
        T target = HystrixFeign.builder()
                .contract(new SpringMvcContract())
                .options(feignConfig.options())
                .logger(new Slf4jLogger())
                .logLevel(feignConfig.level())
                .client(
                        new LoadBalancerFeignClient
                                (new Client.Default(null, null), cachingFactory, clientFactory)
                )
                .decoder(new ResponseEntityDecoder(new SpringDecoder(messageConverters)))
                .encoder(new SpringEncoder(messageConverters))
                .errorDecoder(customErrorDecoder)
                .target(custom); // failed in this method
        return target;
    }
}
```

--- a/core/src/main/java/feign/Contract.java
+++ b/core/src/main/java/feign/Contract.java
@@ -138,11 +138,34 @@ public interface Contract {
     }
 
     private static void checkMapKeys(String name, Type genericType) {
-      Type[] parameterTypes = ((ParameterizedType) genericType).getActualTypeArguments();
-      Class<?> keyClass = (Class<?>) parameterTypes[0];
-      checkState(String.class.equals(keyClass),
-              "%s key must be a String: %s", name, keyClass.getSimpleName());
+      Class<?> keyClass = null;
+
+      // assume our type parameterized
+      if (ParameterizedType.class.isAssignableFrom(genericType.getClass())) {
+        Type[] parameterTypes = ((ParameterizedType) genericType).getActualTypeArguments();
+        keyClass = (Class<?>) parameterTypes[0];
+      } else if (genericType instanceof Class<?>) {
+        // raw class, type parameters cannot be inferred directly, but we can scan any extended
+        // interfaces looking for any explict types
+        Type[] interfaces = ((Class) genericType).getGenericInterfaces();
+        if (interfaces != null) {
+          for (Type extended : interfaces) {
+            if (ParameterizedType.class.isAssignableFrom(extended.getClass())) {
+              // use the first extended interface we find.
+              Type[] parameterTypes = ((ParameterizedType) extended).getActualTypeArguments();
+              keyClass = (Class<?>) parameterTypes[0];
+              break;
+            }
+          }
+        }
+      }
+
+      if (keyClass != null) {
+        checkState(String.class.equals(keyClass),
+            "%s key must be a String: %s", name, keyClass.getSimpleName());
+      }
     }
+
 
     /**
      * Called by parseAndValidateMetadata twice, first on the declaring class, then on the

--- a/core/src/test/java/feign/DefaultContractTest.java
+++ b/core/src/test/java/feign/DefaultContractTest.java
@@ -371,6 +371,12 @@ public class DefaultContractTest {
     }
   }
 
+  @Test
+  public void headerMapSubclass() throws Exception {
+    MethodMetadata md = parseAndValidateMetadata(HeaderMapInterface.class, "headerMapSubClass", SubClassHeaders.class);
+    assertThat(md.headerMapIndex()).isEqualTo(0);
+  }
+
   interface Methods {
 
     @RequestLine("POST /")
@@ -470,6 +476,9 @@ public class DefaultContractTest {
 
     @RequestLine("POST /")
     void multipleHeaderMap(@HeaderMap Map<String, String> headers, @HeaderMap Map<String,String> queries);
+
+    @RequestLine("POST /")
+    void headerMapSubClass(@HeaderMap SubClassHeaders httpHeaders);
   }
 
   interface HeaderParams {
@@ -626,6 +635,12 @@ public class DefaultContractTest {
 
     private List<Entity<K, M>> entities;
   }
+
+
+  interface SubClassHeaders extends Map<String, String> {
+
+  }
+
 
   @Headers("Version: 1")
   interface ParameterizedApi extends ParameterizedBaseApi<String, Long> {

